src/PurchaseForm.elm

```diff
@@ -183,122 +159,122 @@ validateForm productId form =
-        |> Codec.field "billingEmail" .billingEmail emailAddressCodec
-        |> Codec.field "country" .country nonemptyStringCodec
-        |> Codec.field "originCity" .originCity nonemptyStringCodec
-        |> Codec.field "primaryModeOfTravel" .primaryModeOfTravel TravelMode.codec
-        |> Codec.field "grantContribution" .grantContribution Codec.int
-        |> Codec.field "sponsorship" .sponsorship (Codec.maybe Codec.string)
-        |> Codec.buildObject
-
-
-nonemptyStringCodec =
-    Codec.andThen
-        (\text ->
-            case String.Nonempty.fromString text of
-                Just nonempty ->
-                    Codec.succeed nonempty
+        attendees =
+            let
+                attendeesValidated =
+                    form.attendees |> List.map validateAttendee
+            in
+            if attendeesValidated |> List.all Helpers.isJust then
+                attendeesValidated |> Helpers.justs |> Ok
 
-                Nothing ->
-                    Codec.fail ("Invalid nonempty string: " ++ text)
-        )
-        String.Nonempty.toString
-        Codec.string
+            else
+                Err "Invalid attendees"
+    in
+    case T4 billingEmail grantContribution sponsorship attendees of
+        T4 (Ok billingEmailOk) (Ok grantContributionOk) (Ok sponsorshipOk) (Ok attendeesOk) ->
+            Just
+                { attendees = attendeesOk
+                , accommodationBookings = form.accommodationBookings
+                , billingEmail = billingEmailOk
+                , grantContribution = grantContributionOk
+                , grantApply = form.grantApply
+                , sponsorship = sponsorshipOk
+                }
+
+        _ ->
+            Nothing
+
+
+validateAttendee : AttendeeForm -> Maybe AttendeeFormValidated
+validateAttendee form =
+    let
+        name =
+            validateName form.name
 
+        emailAddress =
+            validateEmailAddress form.email
 
-emailAddressCodec : Codec EmailAddress
-emailAddressCodec =
-    Codec.andThen
-        (\text ->
-            case EmailAddress.fromString text of
-                Just email ->
-                    Codec.succeed email
+        country =
+            String.Nonempty.fromString form.country
 
-                Nothing ->
-                    Codec.fail ("Invalid email: " ++ text)
-        )
-        EmailAddress.toString
-        Codec.string
+        originCity =
+            String.Nonempty.fromString form.originCity
+    in
+    case T4 name emailAddress country originCity of
+        T4 (Ok nameOk) (Ok emailAddressOk) (Just countryOk) (Just originCityOk) ->
+            Just
+                { name = nameOk
+                , email = emailAddressOk
+                , country = countryOk
+                , originCity = originCityOk
+                , primaryModeOfTravel = form.primaryModeOfTravel
+                }
+
+        _ ->
+            Nothing
+
+
+
+-- codec : Codec PurchaseFormValidated
```

Comment:
Don't need this to go live?

Reply:
Nope, for the admin display, leaving for later if we want to restore.